### ISSUE TITLE
MultiPartFormData with Custom Endpoint Issue

### ISSUE THREAD

**Issue by syegin** [Moya/Moya#1137]
I've noticed a bug in my project and using Moya 8.0.5.

In endpoint closure, I'm adding some additional parameters and request tasks working as expected. However, one of my task which is an upload multipart, missing these additional parameters. After some debugging, i figured out that:

```swift
if let parameters = target.parameters {
                parameters
                    .flatMap { key, value in multipartQueryComponents(key, value) }
                    .forEach { key, value in
                        if let data = value.data(using: .utf8, allowLossyConversion: false) {
                            form.append(data, withName: key)
                        }
                }
            }
```
In sendUploadMultipart function of MoyaProvider+Internal, formatdata is appending 'target.parameters', not 'endpoint.parameters'. By passing endpoint to function, the problem goes away.

**Comment by BasThomas** [comment]
Can someone confirm this as a bug?

If so, @syegin, would you be willing to do a PR to fix this? :)

**Comment by sunshinejr** [comment]
Hey @syegin, thank you for the issue. Seems like what you're saying might be the case - upload doesn't take parameters, just data, so in case of an `Endpoint`, we just add parameters and not append the data. Was trying to think of a fix and it seems like adding parameters from endpoint to the multipart data [here](https://github.com/Moya/Moya/blob/master/Sources/Moya/MoyaProvider%2BInternal.swift#L90-L92) might be sufficient. Also, we would need to add a test that covers this situation (probably asserting the data). 

Let us know if you want to take a stab at it, @syegin - we will try to help you in any way we can :)

**Comment by syegin** [comment]
Btw, It'is a similar issue with #1066

@sunshinejr, it seems we can't add parameters directly to multipartformdata array just because it's a let constant. But we can generate a new multipart array adding endpoint's parameters with an array extension or a function. Maybe this:

```swift
private extension Array where Element == MultipartFormData {
    
    func appendParameters(_ parameters: [String:Any]?) -> [MultipartFormData] {
        var formData = self
        if let parameters = parameters {
            parameters
                .flatMap { key, value in multipartQueryComponents(key, value) }
                .forEach { key, value in
                    if let data = value.data(using: .utf8, allowLossyConversion: false) {
                        formData.append(MultipartFormData.init(provider: .data(data), name: key))
                    }
            }
        }
        return formData
    }
}
```

And our function call turns to this: 

```swift
cancellableToken.innerCancellable = self.sendUploadMultipart(target, request: preparedRequest, queue: queue, multipartBody: multipartBody.appendParameters(endpoint.parameters), progress: progress, completion: networkCompletion)
```

And finally we need to delete the part that adding target.parameters in sendUploadMultipart function. I've also tested it and worked in my project. I can make these changes you wish, no problem. Thanks for this great library btw :)

**Comment by sunshinejr** [comment]
Now that I think about it, this fix will now transition from ignoring `parameters` in multipart upload to adding them to the request as well. But in normal upload `parameters` will still be ignored. This makes a lot of kinda-undefined behavior. But that's another problem (and I will open a new issue for that). I think we may accept it as a bug fix and release a 8.0.6 for now if needed.

About this changes - seems like a way to go! Please make a PR so we can discuss the implementation details 👍

**Comment by sunshinejr** [comment]
This is being worked on at #1139, so I'm closing this one.
